File: project-{topic}-optimized/src/controllers/accountController.js
const accountService = require('../services/accountService');
const { ApiError } = require('../middleware/errorHandler');
const logger = require('../utils/logger');
const { USER_ROLES } = require('../utils/constants');

/**
 * Creates a new account for an authenticated user.
 * @param {object} req - Express request object.
 * @param {object} res - Express response object.
 * @param {function} next - Express next middleware function.
 */
const createAccount = async (req, res, next) => {
  try {
    const userId = req.user.id; // Get user ID from authenticated request
    const newAccount = await accountService.createAccount(userId, req.body);
    res.status(201).json({
      message: 'Account created successfully',
      account: newAccount,
    });
  } catch (error) {
    logger.error('Error in accountController.createAccount:', error);
    next(error);
  }
};

/**
 * Retrieves an account by ID. Users can only access their own accounts. Admins can access any.
 * @param {object} req - Express request object.
 * @param {object} res - Express response object.
 * @param {function} next - Express next middleware function.
 */
const getAccount = async (req, res, next) => {
  try {
    const accountId = req.params.id;
    const account = await accountService.getAccountById(accountId);

    // Authorization check: User can only view their own account
    if (req.user.role !== USER_ROLES.ADMIN && account.user_id !== req.user.id) {
      return next(new ApiError(403, 'Forbidden: You do not have access to this account.'));
    }

    res.status(200).json({ account });
  } catch (error) {
    logger.error(`Error in accountController.getAccount for ID ${req.params.id}:`, error);
    next(error);
  }
};

/**
 * Retrieves all accounts for the authenticated user.
 * @param {object} req - Express request object.
 * @param {object} res - Express response object.
 * @param {function} next - Express next middleware function.
 */
const getMyAccounts = async (req, res, next) => {
  try {
    const userId = req.user.id;
    const accounts = await accountService.getAccountsByUserId(userId);
    res.status(200).json({ accounts });
  } catch (error) {
    logger.error(`Error in accountController.getMyAccounts for user ${req.user.id}:`, error);
    next(error);
  }
};

/**
 * Updates an account. Users can only update their own accounts (if applicable, limited fields). Admins can update any.
 * Note: Balance updates are specifically handled by the transaction service.
 * @param {object} req - Express request object.
 * @param {object} res - Express response object.
 * @param {function} next - Express next middleware function.
 */
const updateAccount = async (req, res, next) => {
  try {
    const accountId = req.params.id;
    const account = await accountService.getAccountById(accountId); // Get account to check ownership

    // Authorization check
    if (req.user.role !== USER_ROLES.ADMIN && account.user_id !== req.user.id) {
      return next(new ApiError(403, 'Forbidden: You do not have permission to update this account.'));
    }

    const updatedAccount = await accountService.updateAccount(accountId, req.body);
    res.status(200).json({
      message: 'Account updated successfully',
      account: updatedAccount,
    });
  } catch (error) {
    logger.error(`Error in accountController.updateAccount for ID ${req.params.id}:`, error);
    next(error);
  }
};

/**
 * Deletes an account. Users can only delete their own accounts. Admins can delete any.
 * Requires careful consideration in a real payment system (e.g., preventing deletion of accounts with funds).
 * @param {object} req - Express request object.
 * @param {object} res - Express response object.
 * @param {function} next - Express next middleware function.
 */
const deleteAccount = async (req, res, next) => {
  try {
    const accountId = req.params.id;
    const account = await accountService.getAccountById(accountId);

    // Authorization check
    if (req.user.role !== USER_ROLES.ADMIN && account.user_id !== req.user.id) {
      return next(new ApiError(403, 'Forbidden: You do not have permission to delete this account.'));
    }

    const message = await accountService.deleteAccount(accountId);
    res.status(200).json({ message });
  } catch (error) {
    logger.error(`Error in accountController.deleteAccount for ID ${req.params.id}:`, error);
    next(error);
  }
};

module.exports = {
  createAccount,
  getAccount,
  getMyAccounts,
  updateAccount,
  deleteAccount,
};

File: project-{topic}-complete/src/controllers/AccountController.cpp
```cpp
#include "AccountController.h"
#include <nlohmann/json.hpp>

namespace PaymentProcessor {
namespace Controllers {

crow::response AccountController::createAccount(const crow::request& req, AuthContext& authContext) {
    LOG_INFO("Received request to create account by user ID {}.", authContext.userId);

    if (!authContext.isAuthenticated) {
        throw Exceptions::UnauthorizedException();
    }
    // Only ADMIN or MERCHANT users can create accounts
    if (authContext.userRole != "ADMIN" && authContext.userRole != "MERCHANT") {
        throw Exceptions::ForbiddenException("Only ADMIN or MERCHANT users can create accounts.");
    }

    auto jsonBody = nlohmann::json::parse(req.body);

    CreateAccountRequestDTO createDto;
    createDto.from_json(jsonBody);

    // For simplicity, we assume the user creating the account is also its owner.
    // In a real system, an ADMIN might create an account for another merchant user.
    // For now, if no userId is provided in DTO, use authenticated user's ID.
    if (createDto.userId == 0) { // Default value for `long long`
        createDto.userId = authContext.userId;
    } else {
        // If an ADMIN is creating an account for another user, additional checks needed
        if (authContext.userRole != "ADMIN" && createDto.userId != authContext.userId) {
            throw Exceptions::ForbiddenException("MERCHANT users cannot create accounts for other users.");
        }
        // An ADMIN could create an account for any user; for simplicity, we allow it if userId is provided.
    }


    Account newAccount = accountService.createAccount(createDto.userId, createDto.name, createDto.currency, createDto.initialBalance);
    LOG_INFO("Account '{}' (ID: {}) created for user ID {}.", newAccount.name, *newAccount.id, newAccount.userId);

    crow::response res(201); // Created
    res.set_header("Content-Type", "application/json");
    res.write(newAccount.toJson().dump());
    return res;
}

crow::response AccountController::getAccount(const crow::request& req, AuthContext& authContext, long long accountId) {
    LOG_INFO("Received request to get account ID {} by user ID {}.", accountId, authContext.userId);

    if (!authContext.isAuthenticated) {
        throw Exceptions::UnauthorizedException();
    }

    auto account_opt = accountService.getAccountById(accountId);
    if (!account_opt.has_value()) {
        throw Exceptions::NotFoundException("Account with ID " + std::to_string(accountId) + " not found.");
    }
    Account account = account_opt.value();

    // Authorization check: Only owner or ADMIN can view account
    if (authContext.userRole != "ADMIN" && account.userId != authContext.userId) {
        throw Exceptions::ForbiddenException("You do not have permission to view this account.");
    }

    LOG_INFO("Account ID {} retrieved for user ID {}.", accountId, authContext.userId);
    crow::response res(200); // OK
    res.set_header("Content-Type", "application/json");
    res.write(account.toJson().dump());
    return res;
}

crow::response AccountController::getMyAccounts(const crow::request& req, AuthContext& authContext) {
    LOG_INFO("Received request to get accounts for authenticated user ID {}.", authContext.userId);

    if (!authContext.isAuthenticated) {
        throw Exceptions::UnauthorizedException();
    }

    // For "my" accounts, the userId is always the authenticated user's ID
    std::vector<Account> accounts = accountService.getAccountsByUserId(authContext.userId);

    // No specific DTO for list response, just return an array
    nlohmann::json j_array = nlohmann::json::array();
    for (const auto& acc : accounts) {
        j_array.push_back(acc.toJson());
    }

    LOG_INFO("Retrieved {} accounts for user ID {}.", accounts.size(), authContext.userId);
    crow::response res(200); // OK
    res.set_header("Content-Type", "application/json");
    res.write(j_array.dump());
    return res;
}

crow::response AccountController::updateAccount(const crow::request& req, AuthContext& authContext, long long accountId) {
    LOG_INFO("Received request to update account ID {} by user ID {}.", accountId, authContext.userId);

    if (!authContext.isAuthenticated) {
        throw Exceptions::UnauthorizedException();
    }

    auto account_opt = accountService.getAccountById(accountId);
    if (!account_opt.has_value()) {
        throw Exceptions::NotFoundException("Account with ID " + std::to_string(accountId) + " not found for update.");
    }
    Account existingAccount = account_opt.value();

    // Authorization check: Only owner or ADMIN can update account
    if (authContext.userRole != "ADMIN" && existingAccount.userId != authContext.userId) {
        throw Exceptions::ForbiddenException("You do not have permission to update this account.");
    }

    auto jsonBody = nlohmann::json::parse(req.body);
    UpdateAccountRequestDTO updateDto;
    updateDto.from_json(jsonBody);

    Account updatedAccount = accountService.updateAccount(accountId, updateDto.name, updateDto.status);
    LOG_INFO("Account ID {} updated successfully by user ID {}.", accountId, authContext.userId);

    crow::response res(200); // OK
    res.set_header("Content-Type", "application/json");
    res.write(updatedAccount.toJson().dump());
    return res;
}

crow::response AccountController::deleteAccount(const crow::request& req, AuthContext& authContext, long long accountId) {
    LOG_INFO("Received request to delete account ID {} by user ID {}.", accountId, authContext.userId);

    if (!authContext.isAuthenticated) {
        throw Exceptions::UnauthorizedException();
    }

    auto account_opt = accountService.getAccountById(accountId);
    if (!account_opt.has_value()) {
        throw Exceptions::NotFoundException("Account with ID " + std::to_string(accountId) + " not found for deletion.");
    }
    Account existingAccount = account_opt.value();

    // Authorization check: Only ADMIN can delete account for others, owner can delete their own.
    if (authContext.userRole != "ADMIN" && existingAccount.userId != authContext.userId) {
        throw Exceptions::ForbiddenException("You do not have permission to delete this account.");
    }

    accountService.deleteAccount(accountId);
    LOG_INFO("Account ID {} deleted successfully by user ID {}.", accountId, authContext.userId);

    crow::response res(204); // No Content
    return res;
}

} // namespace Controllers
} // namespace PaymentProcessor
```